### PR TITLE
feat: add support for path matching precedence

### LINKED ISSUE #1770: Corner case in HTTPRouteRule match order when RegularExpression is used
**What would you like to be added**:

We need to add test cases to ensure the RegularExpression match happens after the other matches.

**Why this is needed**:

According to the HTTPRouteRule definition:
https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1beta1.HTTPRouteRule

> Proxy or Load Balancer routing configuration generated from HTTPRoutes MUST prioritize matches based on the following criteria, continuing on ties. Across all rules specified on applicable Routes, precedence must be given to the match with the largest number of:
>
>    Characters in a matching path.
>    Header matches.
>    Query param matches.

The routes are sorted by the "Characters in a matching path." first.

However, when the RegularExpression in https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.PathMatchType is involved, the order is ambiguous as the regex can be very specific.

For example, a regex `/a/(.*)+` has more characters in a matching path than exact match `/a/b`, while it is similar to the path prefix `/a/`. If we strictly sort the route by the path, `/a/(.*)+` will be matched before `/a/b`.

An idea to sort this problem is to ensure the regex match happens after the other matches.

### PR DESCRIPTION
**Pull request by Xunzhuo**
**What type of PR is this?**

/kind bug
/area conformance

**What this PR does / why we need it**:

Path matches in following orders:
1. Exact
2. Prefix

**Which issue(s) this PR fixes**:

Fixes #1770

**Does this PR introduce a user-facing change?**:

```release-note
* Updated spec to clarify that Exact matches have precedence over Prefix matches and RegularExpression matches have implementation specific precedence.
* Added conformance test to verify that path matching precedence is implemented.
```

### CODE REVIEW

File: conformance/tests/httproute-path-match-order.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  - matches:
+    - path:
+        type: Exact
+        value: /a/b
+    backendRefs:
+    - name: infra-backend-v1
+      port: 8080
+  - matches:
+    - path:
+        type: RegularExpression
```
Comment (sunjayBhatia) [inline conformance/tests/httproute-path-match-order.yaml:19]: regex match is implementation-specific support so we probably can't use that here: https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.PathMatchType

File: apis/v1beta1/httproute_types.go [hunk 2]
```diff
@@ -156,7 +156,9 @@ type HTTPRouteRule struct {
 	// ties. Across all rules specified on applicable Routes, precedence must be
 	// given to the match with the largest number of:
 	//
-	// * Characters in a matching path.
+	// * Characters in a matching "Exact" path match
+	// * Characters in a matching "RegularExpression" path match
```
Comment (howardjohn) [inline apis/v1beta1/httproute_types.go:160]: Regex between Exact and Prefix feels a bit odd...

TBH, using Characters at all in regex feels odd entirely. Characters has no meaning in regex. is there any proxy implementation that does this today? I worry we will see users inserting dummy characters

Comment (sunjayBhatia) [reply apis/v1beta1/httproute_types.go:160]: contour uses regex between exact and prefix because a regex "could be" more specific than a prefix match, not a perfect heuristic but yeah

also yeah we do sort our regex matches on length just to enable consistent ordering, again not a perfect heuristic

Comment (robscott) [reply apis/v1beta1/httproute_types.go:160]: Chatting with @briantkennedy about this and he pointed out that this would likely not be compatible with NGINX where RegEx matches always happen last. It's probably best to require "Exact", "Prefix", then "RegEx". @brianehlert @pleshakov are probably best to confirm from NGINX side though.

File: conformance/tests/httproute-path-match-order.yaml (hunk 1 above)
Comment (Xunzhuo) [reply conformance/tests/httproute-path-match-order.yaml:19]: Thanks

File: apis/v1beta1/httproute_types.go (hunk 2 above)
Comment (Xunzhuo) [reply apis/v1beta1/httproute_types.go:160]: I am not quite sure for putting the order of regex to where, since it is different in implementations. I just added conformance to test:
```
Characters in a matching "Exact" path match
Characters in a matching "Prefix" path match
```

File: conformance/tests/httproute-path-match-order.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+)
+
+func init() {
+	ConformanceTests = append(ConformanceTests, HTTPRoutePathMatchOrders)
+}
+
+var HTTPRoutePathMatchOrders = suite.ConformanceTest{
+	ShortName:   "HTTPRoutePathMatchOrders",
+	Description: "A HTTPRoute with different path matches for different backends",
+	Manifests:   []string{"tests/httproute-path-match-orders.yaml"},
```
Comment (sunjayBhatia) [inline conformance/tests/httproute-path-match-order.go:36]: typo here in the test yaml file name

**Review by sunjayBhatia — CHANGES_REQUESTED** [review]

File: conformance/tests/httproute-path-match-order.go (hunk 3 above)
Comment (sunjayBhatia) [reply conformance/tests/httproute-path-match-order.go:36]: would suggest making sure "Orders/orders" -> "Order/order"

File: apis/v1beta1/httproute_types.go (hunk 2 above)
Comment (robscott) [reply apis/v1beta1/httproute_types.go:160]: I don't think we're going to be able to mandate any specific precedence for RegEx path matches in the API. Instead, we probably need to label that as implementation-specific. Looking at the [NGINX reference docs], it does not look like it would be possible for it to be conformant with any precedence based on the length of a regex.

>To find location matching a given request, nginx first checks locations defined using the prefix strings (prefix locations). Among them, the location with the longest matching prefix is selected and remembered. Then regular expressions are checked, in the order of their appearance in the configuration file. The search of regular expressions terminates on the first match, and the corresponding configuration is used. If no match with a regular expression is found then the configuration of the prefix location remembered earlier is used.

Since we already consider RegEx path matching "Implementation-Specific" we likely should also consider the precedence of RegEx path matches "Implementation-Specific".

I'd recommend removing the "RegEx" mention from the list and adding a note below the list that states that "The precedence of RegularExpression path matches are implementation-specific."

Comment (Xunzhuo) [reply apis/v1beta1/httproute_types.go:160]: Thanks @robscott

File: conformance/tests/httproute-path-match-order.go (hunk 3 above)
Comment (Xunzhuo) [reply conformance/tests/httproute-path-match-order.go:36]: Thanks @sunjayBhatia !

**Review by sunjayBhatia — APPROVED** [review]
lgtm, found a little bug in contour as part of running this https://github.com/projectcontour/contour/pull/5217

nice!

**Review by sunjayBhatia — APPROVED** [review]

File: apis/v1beta1/httproute_types.go [hunk 4]
```diff
@@ -156,10 +156,13 @@ type HTTPRouteRule struct {
 	// ties. Across all rules specified on applicable Routes, precedence must be
 	// given to the match with the largest number of:
 	//
-	// * Characters in a matching path.
+	// * Characters in a matching "Exact" path match
+	// * Characters in a matching "Prefix" path match
 	// * Header matches.
 	// * Query param matches.
 	//
+	// Note: The precedence of RegularExpression path matches are implementation-specific.
```
Comment (shaneutt) [inline apis/v1beta1/httproute_types.go:164]: I agree with the decision to make precedence for RE implementation-specific, as I think we would likely have created a conundrum for several implementations.

File: conformance/tests/httproute-path-match-order.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+)
+
+func init() {
+	ConformanceTests = append(ConformanceTests, HTTPRoutePathMatchOrder)
+}
+
+var HTTPRoutePathMatchOrder = suite.ConformanceTest{
+	ShortName:   "HTTPRoutePathMatchOrder",
+	Description: "A HTTPRoute with different path matches for different backends",
```
Comment (shaneutt) [inline conformance/tests/httproute-path-match-order.go:35]: Am i understanding the intention of this test to be something more like:

```suggestion
	Description: "An HTTPRoute can route multiple separate path matches to a single backend",
```

?

Comment (shaneutt) [reply conformance/tests/httproute-path-match-order.go:35]: Made an edit above, lmkwyt

File: conformance/tests/httproute-path-match-order.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (robscott) [inline conformance/tests/httproute-path-match-order.go:2]:
```suggestion
Copyright 2023 The Kubernetes Authors.
```

File: conformance/tests/httproute-path-match-order.yaml [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    - path:
+        type: PathPrefix
+        value: /match/prefix/
+    backendRefs:
+    - name: infra-backend-v1
+      port: 8080
+  - matches:
+    - path:
+        type: PathPrefix
+        value: /match/prefix/one/
```
Comment (robscott) [inline conformance/tests/httproute-path-match-order.yaml:48]: It may be helpful to have at least one of these prefix matches without a trailing slash.

**Review by robscott — COMMENTED** [review]
This is really great, thanks @Xunzhuo! A couple tiny nits but otherwise LGTM.

File: conformance/tests/httproute-path-match-order.go (hunk 5 above)
Comment (Xunzhuo) [reply conformance/tests/httproute-path-match-order.go]: Thanks @shaneutt ans @sunjayBhatia !

File: conformance/tests/httproute-path-match-order.yaml (hunk 7 above)
Comment (Xunzhuo) [reply conformance/tests/httproute-path-match-order.yaml]: Done Thanks!

**Comment by robscott** [comment]
Thanks @Xunzhuo! Will wait for someone else to LGTM in case I missed anything in the conformance tests.

/cc @arkodg @mlavacca 
/approve

**Comment by arkodg** [comment]
hey @Xunzhuo have you tested this test against any implementation ?

**Comment by Xunzhuo** [comment]
@sunjayBhatia tested it in Contour, I will have a test in EG.

**Review by arkodg — APPROVED** [review]
